djcurlz/deposits-exercise
# feat: new rails 7 app to implement deposits exercise in

# Deposits Exercise

## Requirements

A tradeline on a credit report represents an account, such as a credit card or car loan.

In the codebase, the following exists:

* The `Tradeline` model, which can be used to store tradelines in the database.
* A `Tradeline` controller, which serves both the `index` and `show` routes for tradelines.
* Basic `rspec` tests on the controller with a `tradeline` FactoryBot.

Add in the ability to create deposits for a specific tradeline using a JSON API call. You should store the deposit date (this is most likely a future date), and amount. Expose the outstanding balance for a given tradeline in the JSON output for `TradelinesController#show`, and `TradelinesController#index`. The oustanding balance should be calculated as the `amount` minus all of the deposit amounts.

In addition:

* A deposit should not be able to be created that exceeds the outstanding balance of a tradeline.
* You should return validation errors if a deposit cannot be created.
* Expose endpoints for seeing all deposits for a given tradeline, as well as seeing an individual deposit.

## Setup

To test this branch locally please ensure the following are installed and setup on your local machine.
- [POSTMAN](https://learning.postman.com/docs/getting-started/installation/installation-and-updates/)
- [RUBY 3.3.1](https://www.ruby-lang.org/en/documentation/installation/)
- [RUBY ON RAILS 7.1](https://mac.install.guide/rubyonrails/)
- [PostgreSQL](https://www.moncefbelyamani.com/how-to-install-postgresql-on-a-mac-with-homebrew-and-lunchy/)

Note: PostgreSQL is optional but what I used to test my solution locally, if you prefer MySQL simply updated the config/database.yml to use that instead.

## Testing

Once you have Postman, Ruby and Ruby on Rails installed run the following steps to test the solution.

1. Clone github repo locally
```yaml
git clone
```

2. Setup database and run migrations
```yaml
rails db:create
```
```yaml
rails db:migrate
```

3. Run controller rspec tests locally
```yaml
rspec -f documentation spec/controllers/tradelines_controller_spec.rb
```
```yaml
rspec -f documentation spec/controllers/deposits_controller_spec.rb
```

4. Start rails server locally with `rails s`

5. Open Postman and import the environment and collection files respectively.
![PostmanFilesToImport](https://github.com/user-attachments/assets/aa67b9a6-c3a0-4268-8225-d5839cc50033)

6.  Run through each of the collection requests in the order shown. 

Observe when executing GET Tradelines request the response includes the outstanding balance.
![TradelinesOutstandingBalances](https://github.com/user-attachments/assets/7495fdea-73b6-42ea-bcb3-7be97cbc7d63)

You should see the validation error triggered for both invalid deposits (Chase, Discover) with the message below.

![InvalidDepositValidation](https://github.com/user-attachments/assets/0076b482-436f-400d-b40a-2cdf61e66778)
